atc/db/pipeline.go

```diff
@@ -1136,35 +1138,40 @@ func (p *pipeline) getBuildsFrom(tx Tx, col string) (map[string]Build, error) {
 // plug the global variables, otherwise just return the global variables.
 func (p *pipeline) Variables(logger lager.Logger, globalSecrets creds.Secrets, varSourcePool creds.VarSourcePool) (vars.Variables, error) {
 	globalVars := creds.NewVariables(globalSecrets, p.TeamName(), p.Name(), false)
-	varss := []vars.Variables{}
-	for _, cm := range p.varSources {
+	varss := []vars.Variables{globalVars}
+
+	orderedVarSources, err := p.varSources.OrderByDependency()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, cm := range orderedVarSources {
 		factory := creds.ManagerFactories()[cm.Type]
 		if factory == nil {
 			return nil, fmt.Errorf("unknown credential manager type: %s", cm.Type)
 		}
 
 		// Interpolate variables in pipeline credential manager's config
-		newConfig, err := creds.NewParams(globalVars, atc.Params{"config": cm.Config}).Evaluate()
+		newConfig, err := creds.NewParams(vars.NewMultiVars(varss), atc.Params{"config": cm.Config}).Evaluate()
 		if err != nil {
-			return nil, err
+			return nil, errors.Wrapf(err, "evaluate var_source '%s' error", cm.Name)
 		}
 
 		config, ok := newConfig["config"].(map[string]interface{})
 		if !ok {
-			return nil, fmt.Errorf("invalid config format")
+			return nil, fmt.Errorf("var_source '%s' invalid config", cm.Name)
 		}
-		secrets, err := varSourcePool.FindOrCreate(logger, config, factory)
+		secrets, err := varSourcePool.FindOrCreate(logger, cm.Name, config, factory)
 		if err != nil {
-			return nil, err
+			return nil, errors.Wrapf(err, "create var_source '%s' error", cm.Name)
 		}
 		varss = append(varss, creds.NewVariables(secrets, p.TeamName(), p.Name(), true))
 	}
 
-	if len(varss) == 0 {
+	if len(varss) == 1 {
 		return globalVars, nil
 	}
 
-	varss = append(varss, globalVars)
 	return vars.NewMultiVars(varss), nil
```

Comment:
What if instead of just listing everything under `vars.NewMultiVars` and having var sources know their own name, we could have our own `vars.NamedVariables` type which does the lookup by-name through a map?

Something like this:

```go
namedVars := vars.NamedVariables{}

allVars := vars.NewMultiVars([]vars.Variables{globalVars, namedVars})

for _, cm := range orderedVarSources {
  // ...

  // allVars here sneakily relies on us mutating namedVars
  // ¯\_(ツ)_/¯ should be safe!
  newConfig, err := creds.NewParams(allVars, atc.Params{"config": cm.Config}).Evaluate()
  if err != nil {
    // ...
  }

  // this doesn't need a name anymore
  secrets, err := varSourcePool.FindOrCreate(logger, config, factory)
  if err != nil {
    // ...
  }

  // ...

  namedVars[cm.Name] = creds.NewVariables(secrets, p.TeamName(), p.Name(), true)
}

return allVars, nil
```

Then `(vars.NamedVariables).Get` type would just do a map lookup and delegate that way.

Reply:
This is a good idea, `vars.NamedVariables` is betters than `creds.NamedSecrets`.